### PR TITLE
POC: Plugin providers

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Introduce a new plugin type "provider" especially for backend plugins to 
allow a better distinction between client side and backend plugins.
Current way of defining an executable of app/data source plugin.json
is still supported. However, by introducing the new provider type
it allows multiple data source, app and panel plugins to add a
provider as dependency and by that allowing one provider to be
shared/used by multiple plugins. Earlier this required one
backend process per each plugin that wanted to make use of a
backend plugin.

In addition, this POC also explores a new backend plugin
capability, Provider, which, if implemented by a plugin,
would be called by the backend plugin manager after the plugin
has started to ask it for a an optional list of providers. This
basically should allow a backend plugin/provider to proxy/bundle
several providers in one binary, even though it's a bit rough
around the corners currently. This idea has currently been 
implemented for the two core data sources that current has 
been converted to use SDK contracts, test data source and 
CloudWatch.

This is basically a POC following the [Plugin providers design doc](https://docs.google.com/document/d/1lbiabeyvhNjN2ACvHr6o55FgcGSVu3WQhhu0J_79n8U/edit#), trying to explain by code rather than text since the subject is a bit hard to explain in words.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/plugins/backendplugin/manager/manager.go [hunk 1]
```diff
@@ -111,6 +114,41 @@ func (m *manager) RegisterAndStart(ctx context.Context, pluginID string, factory
...
+	if err != nil && !errors.Is(err, backendplugin.ErrMethodNotImplemented) {
+		if stopErr := m.UnregisterAndStop(ctx, pluginID); stopErr != nil {
+			p.Logger().Error("Failed to unregister and stop plugin", "error", stopErr)
+		}
+		return err
+	}
+
+	m.pluginsMu.Lock()
+
+	for plugin, provider := range m.providerForPlugin {
```
Comment (marefr) [inline pkg/plugins/backendplugin/manager/manager.go:127]: Current hack for re-mapping provider to support multiple plugins.

File: pkg/plugins/backendplugin/manager/manager.go [hunk 2]
```diff
@@ -148,10 +193,23 @@ func (m *manager) IsRegistered(pluginID string) bool {
 
 func (m *manager) Get(pluginID string) (backendplugin.Plugin, bool) {
 	m.pluginsMu.RLock()
-	p, ok := m.plugins[pluginID]
+	p, exists := m.plugins[pluginID]
+	if !exists {
+		if pDependency, providerForPluginExists := m.providerForPluginOverride[pluginID]; providerForPluginExists {
```
Comment (marefr) [inline pkg/plugins/backendplugin/manager/manager.go:198]: Current hack to fallback to re-mapped plugin/provider

File: pkg/plugins/backendplugin/manager/manager.go [hunk 3]
```diff
@@ -244,14 +285,15 @@ func (m *manager) stop(ctx context.Context) {
...
-	m.pluginsMu.RLock()
-	p, registered := m.plugins[pluginID]
-	m.pluginsMu.RUnlock()
-
+	p, registered := m.Get(pluginID)
 	if !registered {
 		return nil, backendplugin.ErrPluginNotRegistered
 	}
 
+	if providerID, exists := m.GetProviderForPlugin(pluginID); exists {
```
Comment (marefr) [inline pkg/plugins/backendplugin/manager/manager.go:293]: Current hack for providing the actual provider and allow a provider to proxy/bundle/support serving multiple providers. This might be a property on `backend.PluginContext` eventually and/or included in context/gRPC metadata to allow proxying provider gRPC requests without unpacking/decode the full gRPC payload.

File: pkg/tsdb/provider_proxy.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	opts, exists := p.providers[providerID]
+	return opts, exists
+}
+
+func (p *providerProxy) CollectMetrics(ctx context.Context) (*backend.CollectMetricsResult, error) {
+	return nil, backendplugin.ErrMethodNotImplemented
+}
+
+func (p *providerProxy) CheckHealth(ctx context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
+	providerID, err := backendplugin.ProviderIDFromContext(ctx)
```
Comment (marefr) [inline pkg/tsdb/provider_proxy.go:50]: See other comment regarding why this is a bit of a hack currently and why we currently need it.

**Comment by ryantxu** [comment]
Is this something we should consider *after* we have a robust plugin manager?  I am convinced that replacing how/where we get plugin information will be a more natural fit than trying to shoe-horn this into existing plugin infrastruture.  ie, cloud could/should get its list of plugins from a query/config not by scanning a file system!

**Comment by marefr** [comment]
@ryantxu Sorry forgot to answer. 

> Is this something we should consider after we have a robust plugin manager?

Robust plugin manager definitely comes first, i.e. @wbrowne ongoing refactoring. 

This PR is a draft and POC and there's currently no plan of merging this. More to receive feedback/get the discussions going by showing some code in addition to problems/proposed solutions in referenced design doc. 

I guess the most interesting part of this POC is that Grafana, when loading a backend plugin, could ask what providers it supports to allow one backend plugin to bundle multiple providers and/or act as a proxy to multiple providers.

**Comment by marefr** [comment]
Closing this for now
